### Analysing ebay car sales

*Business Problem: Cleaning and analysing the cleaned used car listings scraped data for a classifieds section of the German ebay website.*

*Source of the data: https://data.world/data-society/used-cars-data*



In [43]:
import pandas as pd
import numpy as np

In [44]:
autos = pd.read_csv('autos.csv', encoding = 'Windows-1252')

In [45]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [46]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [47]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

*Key Observations*

1. There are null values for the columns:(vehicleType,gearbox,model,fuelType,notRepairedDamage)

2. There are ints and object type columns in our dataset, which might need to be altered, depending upon the analysis.

3. The first few rows also show that there are multiple date columns, out of which the first one shows the date of crawling, one shows the last seen timings and a column that shows the creation of listing.

4. In addition, there are columns with German and English strings, whih might need to be replaced or combined, depending upon our analysis.

5. A few columns are in camelca and a few are in snakecase, which complicates the analysis a little bit.

In [48]:
autos_copy=autos.columns.copy()

In [49]:
def clean_column_names(s):
    s=s.replace("yearOfRegistration","registration_year")
    s=s.replace("monthOfRegistration","registration_month")
    s=s.replace("notRepairedDamage","unrepaired_damage")
    s=s.replace("dateCreated","ad_created")
    s=s.lower()
    return s

names = []

for name in autos_copy:
    name = clean_column_names(name)
    names.append(name)
autos.columns = names  

In [50]:
autos.columns

Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'kilometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')

**Some of the column names still need to be converted from camelcase to snakecase**

In [51]:
autos.head()

datecrawled                            name  seller offertype  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicletype  registration_year    gearbox  powerps  model  \
0    480   test         NaN               1993    manuell        0   golf   
1  18300   test       coupe               2011    manuell      190    NaN   
2   9800   test         suv               2004  automatik      163  grand   
3   1500   test  kleinwagen               2001    manuell       75   golf   
4   3600   test  kleinwagen               2008    manuell       69  fabia   

   kilometer  registration_month fueltype       brand unrepaired_damage  \
0     150000                   0   benzin  volkswagen               NaN   
1     125000                   5   diesel        audi                ja   
2     125000                   8   diesel        jeep               NaN   
3     150000                   6   benzin  volkswagen              nein   
4      90000                   7   diesel       skoda              nein   

            ad_created  nrofpictures  postalcode             lastseen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

**Made changes to column names to more human readable names, which will make the analysis easier as move forward into the project**

In [52]:
autos.describe(include = 'all')

datecrawled         name  seller offertype         price  \
count                371528       371528  371528    371528  3.715280e+05   
unique               280500       233531       2         2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat   Angebot           NaN   
freq                      7          657  371525    371516           NaN   
mean                    NaN          NaN     NaN       NaN  1.729514e+04   
std                     NaN          NaN     NaN       NaN  3.587954e+06   
min                     NaN          NaN     NaN       NaN  0.000000e+00   
25%                     NaN          NaN     NaN       NaN  1.150000e+03   
50%                     NaN          NaN     NaN       NaN  2.950000e+03   
75%                     NaN          NaN     NaN       NaN  7.200000e+03   
max                     NaN          NaN     NaN       NaN  2.147484e+09   

        abtest vehicletype  registration_year  gearbox        powerps   model  \
count   371528      333659      371528.000000   351319  371528.000000  351044   
unique       2           8                NaN        2            NaN     251   
top       test   limousine                NaN  manuell            NaN    golf   
freq    192585       95894                NaN   274214            NaN   30070   
mean       NaN         NaN        2004.577997      NaN     115.549477     NaN   
std        NaN         NaN          92.866598      NaN     192.139578     NaN   
min        NaN         NaN        1000.000000      NaN       0.000000     NaN   
25%        NaN         NaN        1999.000000      NaN      70.000000     NaN   
50%        NaN         NaN        2003.000000      NaN     105.000000     NaN   
75%        NaN         NaN        2008.000000      NaN     150.000000     NaN   
max        NaN         NaN        9999.000000      NaN   20000.000000     NaN   

            kilometer  registration_month fueltype       brand  \
count   371528.000000       371528.000000   338142      371528   
unique            NaN                 NaN        7          40   
top               NaN                 NaN   benzin  volkswagen   
freq              NaN                 NaN   223857       79640   
mean    125618.688228            5.734445      NaN         NaN   
std      40112.337051            3.712412      NaN         NaN   
min       5000.000000            0.000000      NaN         NaN   
25%     125000.000000            3.000000      NaN         NaN   
50%     150000.000000            6.000000      NaN         NaN   
75%     150000.000000            9.000000      NaN         NaN   
max     150000.000000           12.000000      NaN         NaN   

       unrepaired_damage           ad_created  nrofpictures    postalcode  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                   lastseen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

**There are a few columns that just have one or two values and are possible candidates for removal later on, for example**:

1. Offertype
2. Seller


**Price and Kilometer columns are text types and will need exploration later on, as well as the "registration_year" column which is not in the correct date format yet.

In [53]:
#autos["price"] = autos["price"].str.replace("$","")
#autos["price"] = autos["price"].str.replace(",","")
#autos["odometer"] = autos["odometer"].str.replace("km","")
#autos["odometer"] = autos["odometer"].str.replace(",","")
    

In [54]:
autos.head()

datecrawled                            name  seller offertype  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicletype  registration_year    gearbox  powerps  model  \
0    480   test         NaN               1993    manuell        0   golf   
1  18300   test       coupe               2011    manuell      190    NaN   
2   9800   test         suv               2004  automatik      163  grand   
3   1500   test  kleinwagen               2001    manuell       75   golf   
4   3600   test  kleinwagen               2008    manuell       69  fabia   

   kilometer  registration_month fueltype       brand unrepaired_damage  \
0     150000                   0   benzin  volkswagen               NaN   
1     125000                   5   diesel        audi                ja   
2     125000                   8   diesel        jeep               NaN   
3     150000                   6   benzin  volkswagen              nein   
4      90000                   7   diesel       skoda              nein   

            ad_created  nrofpictures  postalcode             lastseen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [55]:
autos["kilometer"] = pd.to_numeric(autos["kilometer"])
autos["price"] = pd.to_numeric(autos["price"])

In [56]:
autos=autos.rename(columns = {"kilometer":"kilometer_km"})

In [57]:
autos["price"].unique().shape

(5597,)

In [58]:
autos["kilometer_km"].unique().shape

(13,)

In [59]:
autos["price"].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [60]:
autos["kilometer_km"].describe()

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer_km, dtype: float64

In [61]:
autos["price"].value_counts(ascending = False).sort_index(ascending = True).head()

0    10778
1     1189
2       12
3        8
4        1
Name: price, dtype: int64

In [62]:
autos["kilometer_km"].value_counts(ascending = True).head()

10000    1949
20000    5676
30000    6041
40000    6376
5000     7069
Name: kilometer_km, dtype: int64

In [63]:
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index()

2014-03-10 00:00:00    0.000003
2015-03-20 00:00:00    0.000003
2015-06-11 00:00:00    0.000003
2015-06-18 00:00:00    0.000003
2015-08-07 00:00:00    0.000003
                         ...   
2016-04-03 00:00:00    0.038893
2016-04-04 00:00:00    0.037741
2016-04-05 00:00:00    0.011655
2016-04-06 00:00:00    0.003155
2016-04-07 00:00:00    0.001558
Name: ad_created, Length: 114, dtype: float64

In [64]:
autos["datecrawled"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:06:22    0.000003
2016-03-05 14:06:23    0.000003
2016-03-05 14:06:24    0.000008
2016-03-05 14:06:25    0.000005
2016-03-05 14:06:26    0.000003
                         ...   
2016-04-07 14:36:54    0.000003
2016-04-07 14:36:55    0.000003
2016-04-07 14:36:56    0.000005
2016-04-07 14:36:57    0.000003
2016-04-07 14:36:58    0.000003
Name: datecrawled, Length: 280500, dtype: float64

In [65]:
autos["lastseen"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:15:08    0.000003
2016-03-05 14:15:16    0.000003
2016-03-05 14:15:39    0.000003
2016-03-05 14:18:30    0.000003
2016-03-05 14:25:59    0.000003
                         ...   
2016-04-07 14:58:47    0.000024
2016-04-07 14:58:48    0.000024
2016-04-07 14:58:49    0.000030
2016-04-07 14:58:50    0.000030
2016-04-07 14:58:51    0.000003
Name: lastseen, Length: 182806, dtype: float64

**Use this cell to explain the findings from the above data exploration**

In [66]:
autos["registration_year"].describe()

count    371528.000000
mean       2004.577997
std          92.866598
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

**The registration year cannot be 1000 (min value) and also cant be 9999 (max value)**. Therefore, will pick a more reasonable range of years for the analysis. 

Will count all the listings of cas within 1900 and 2016 and see if that starts to make more sense.

In [67]:
autos[autos["registration_year"].between(1900,2016)].describe()

price  registration_year        powerps   kilometer_km  \
count  3.567800e+05      356780.000000  356780.000000  356780.000000   
mean   1.742357e+04        2002.779805     116.608493  125416.405068   
std    3.657289e+06           7.395775     189.427653   40158.087893   
min    0.000000e+00        1910.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      71.000000  100000.000000   
50%    2.999000e+03        2003.000000     106.000000  150000.000000   
75%    7.400000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        2016.000000   20000.000000  150000.000000   

       registration_month  nrofpictures     postalcode  
count       356780.000000      356780.0  356780.000000  
mean             5.775133           0.0   50925.143553  
std              3.694198           0.0   25820.697634  
min              0.000000           0.0    1067.000000  
25%              3.000000           0.0   30539.000000  
50%              6.000000           0.0   49716.000000  
75%              9.000000           0.0   71665.000000  
max             12.000000           0.0   99998.000000

In [68]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [69]:
autos.describe()

price  registration_year        powerps   kilometer_km  \
count  3.567800e+05      356780.000000  356780.000000  356780.000000   
mean   1.742357e+04        2002.779805     116.608493  125416.405068   
std    3.657289e+06           7.395775     189.427653   40158.087893   
min    0.000000e+00        1910.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      71.000000  100000.000000   
50%    2.999000e+03        2003.000000     106.000000  150000.000000   
75%    7.400000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        2016.000000   20000.000000  150000.000000   

       registration_month  nrofpictures     postalcode  
count       356780.000000      356780.0  356780.000000  
mean             5.775133           0.0   50925.143553  
std              3.694198           0.0   25820.697634  
min              0.000000           0.0    1067.000000  
25%              3.000000           0.0   30539.000000  
50%              6.000000           0.0   49716.000000  
75%              9.000000           0.0   71665.000000  
max             12.000000           0.0   99998.000000

In [70]:
autos["registration_year"].value_counts(normalize = True)

2000    0.068813
1999    0.063812
2005    0.062548
2006    0.056702
2001    0.056668
          ...   
1925    0.000003
1920    0.000003
1919    0.000003
1915    0.000003
1911    0.000003
Name: registration_year, Length: 97, dtype: float64

**Write a summary of the above steps, as in the removal of outliers and the calculation of value counts for the registration year.**

In [71]:
count_of_brands = autos["brand"].value_counts(normalize=True)

*Restricting the above analysis to the top 5% of the brands, would be a reasonable assumption here*

In [72]:
most_common_brands = count_of_brands[count_of_brands>0.05].index

In [73]:
most_common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [74]:
aggregation_over_brands = {}

for b in most_common_brands:
    new_brands = autos[autos["brand"]==b]
    mean_price = new_brands["price"].mean()
    aggregation_over_brands[b]=int(mean_price)

In [75]:
aggregation_over_brands

{'volkswagen': 13643,
 'bmw': 14798,
 'opel': 3248,
 'mercedes_benz': 17614,
 'audi': 16218,
 'ford': 8702}

**Explain your results here**

In [77]:
mean_mileage_dict = {}

for b in most_common_brands:
    new_brands = autos[autos["brand"]==b]
    mean_mileage = new_brands["kilometer_km"].mean()
    mean_mileage_dict[b] = mean_mileage

In [78]:
mean_mileage_dict

{'volkswagen': 128337.16035208573,
 'bmw': 132657.07642384444,
 'opel': 128755.89005235602,
 'mercedes_benz': 130580.6922739886,
 'audi': 129491.20110417517,
 'ford': 123618.09965803615}

In [79]:
mean_price_s=pd.Series(aggregation_over_brands)

In [80]:
print(mean_price_s)
df_price=pd.DataFrame(mean_price_s, columns = ['mean_price'])
df_price

volkswagen       13643
bmw              14798
opel              3248
mercedes_benz    17614
audi             16218
ford              8702
dtype: int64


mean_price
volkswagen          13643
bmw                 14798
opel                 3248
mercedes_benz       17614
audi                16218
ford                 8702

In [81]:
mean_mileage_s = pd.Series(mean_mileage_dict)
print(mean_mileage_s)

volkswagen       128337.160352
bmw              132657.076424
opel             128755.890052
mercedes_benz    130580.692274
audi             129491.201104
ford             123618.099658
dtype: float64


In [82]:
df_mileage = pd.DataFrame(mean_mileage_s, columns = ['mileage'])
df_mileage

mileage
volkswagen     128337.160352
bmw            132657.076424
opel           128755.890052
mercedes_benz  130580.692274
audi           129491.201104
ford           123618.099658